### PR TITLE
feat: Discord bot

### PR DESCRIPTION
**Pull request by rsrbk**
## Description

Adds a new class `DiscordBot` that creates a bot for a specific discord server, which is fully powered by the underlying `ChatAgent` object.

## Motivation and Context

We want users of CAMEL framework to set up discord and other types of bots in one click.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/bots/discord_bot.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    """
+
+    def __init__(
+        self,
+        chat_agent: ChatAgent,
+        channel_ids: Optional[List[int]] = None,
+        discord_token: Optional[str] = None,
+    ) -> None:
+        self.chat_agent = chat_agent
+        self.token = discord_token or os.getenv('DISCORD_TOKEN')
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:41]: move check for `self.token` into __init__?

File: camel/bots/discord_bot.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    async def on_message(self, message: 'Message') -> None:
+        r"""Event handler for when a message is received.
+
+        Parameters:
+        - message (discord.Message): The message object received.
+
+        Returns:
+        - None
+
+        """
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:85]:
```suggestion
        """
```

File: camel/bots/discord_bot.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        r"""This method is called when the bot has successfully connected to the Discord server.
+        It prints a message indicating that the bot has logged in and displays the username of the bot.
+        """
+        print(f'We have logged in as {self.client.user}')
+
+    async def on_message(self, message: 'Message') -> None:
+        r"""Event handler for when a message is received.
+
+        Parameters:
+        - message (discord.Message): The message object received.
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:80]:
```suggestion
        Args:
            message (discord.Message): The message object received.
```

File: camel/bots/discord_bot.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        if not self.token:
+            raise ValueError(
+                "`DISCORD_TOKEN` not found in environment variables. Get it here: `https://discord.com/developers/applications`."
+            )
+        self.client.run(self.token)
+
+    async def on_ready(self) -> None:
+        r"""This method is called when the bot has successfully connected to the Discord server.
+        It prints a message indicating that the bot has logged in and displays the username of the bot.
+        """
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:73]:
```suggestion
        r"""This method is called when the bot has successfully connected to the Discord server.

        It prints a message indicating that the bot has logged in and displays the username of the bot.
        """
```

File: camel/bots/discord_bot.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        Returns:
+        - None
+
+        """
+        if message.author == self.client.user:
+            return
+
+        if self.channel_ids is not None:
+            if message.channel.id not in self.channel_ids:
+                return
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:91]:
```suggestion
        if self.channel_ids and message.channel.id not in self.channel_ids:
            return
```

File: camel/bots/discord_bot.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            )
+        self.client.run(self.token)
+
+    async def on_ready(self) -> None:
+        r"""This method is called when the bot has successfully connected to the Discord server.
+        It prints a message indicating that the bot has logged in and displays the username of the bot.
+        """
+        print(f'We have logged in as {self.client.user}')
+
+    async def on_message(self, message: 'Message') -> None:
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:76]:
```suggestion
    async def on_message(self, message: Message) -> None:
```

File: camel/bots/discord_bot.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import os
+from typing import TYPE_CHECKING, List, Optional
+
+from camel.agents import ChatAgent
+from camel.messages import BaseMessage
+
+# Conditionally import telebot types only for type checking
+if TYPE_CHECKING:
+    from discord import Message  # type: ignore[import-untyped]
```
Comment (Wendong-Fan) [inline camel/bots/discord_bot.py:22]: seems not necessary to ignore the type check here

**Review by Wendong-Fan — APPROVED** [review]
Thanks for the PR, left some comments related to format, over all LGTM

**Review by zechengz — APPROVED** [review]

File: camel/bots/discord_bot.py (hunk 4 above)
Comment (willshang76) [reply camel/bots/discord_bot.py:73]: Done

File: camel/bots/discord_bot.py (hunk 7 above)
Comment (willshang76) [reply camel/bots/discord_bot.py:22]: Removed

File: camel/bots/discord_bot.py (hunk 1 above)
Comment (willshang76) [reply camel/bots/discord_bot.py:37]: Could you explain more? The code is already in the init function.

Comment (Wendong-Fan) [reply camel/bots/discord_bot.py:37]: in `def run(self) -> None:` we check the token, maybe better to move that to here

**Review by willshang76 — APPROVED** [review]

**Pull request merged**